src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetCoreMvc2Integration.cs

```diff
@@ -0,0 +1,319 @@
+                Log.Error($"Exception when calling {methodDef}.", ex);
+            }
+            finally
+            {
+                aspNetCoreMvcActionScope?.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Wrapper method used to catch unhandled exceptions in the incoming request pipeline for Microsoft.AspNetCore.Mvc.Core
+        /// </summary>
+        /// <param name="context">The DiagnosticSource that this extension method was called on.</param>
+        /// <param name="opCode">The OpCode used in the original method call.</param>
+        /// <param name="mdToken">The mdToken of the original method call.</param>
+        [InterceptMethod(
+            CallerAssembly = AspnetMvcCore,
+            TargetAssembly = AspnetMvcCore,
+            TargetType = ResourceInvoker,
+            TargetSignatureTypes = new[] { ClrNames.Void, ClrNames.Ignore },
+            TargetMinimumVersion = Major2,
+            TargetMaximumVersion = Major2)]
+        public static void Rethrow(object context, int opCode, int mdToken)
+        {
+            string methodDef = $"{ResourceInvoker}.{nameof(Rethrow)}({context?.GetType().FullName} context)";
+            var shouldTrace = Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName);
+
+            Action<object> instrumentedMethod = null;
+
+            try
+            {
+                instrumentedMethod =
+                    MethodBuilder<Action<object>>
+                       .Start(Assembly.GetCallingAssembly(), mdToken, opCode)
+                       .WithConcreteTypeName(ResourceInvoker)
+                       .WithParameters(context)
+                       .Build();
+            }
+            catch (Exception ex)
+            {
+                // profiled app will not continue working as expected without this method
+                Log.ErrorException($"Error retrieving {methodDef}", ex);
+                throw;
+            }
+
+            AspNetAmbientContext ambientContext = null;
+            var exceptionToGrab = context.GetProperty<Exception>("Exception");
+
+            if (shouldTrace)
+            {
+                var httpContextResult = context.GetProperty("HttpContext");
+
+                if (httpContextResult.HasValue)
+                {
+                    ambientContext = AspNetAmbientContext.RetrieveFromHttpContext(httpContextResult.Value);
+                }
+
+                if (ambientContext == null)
+                {
+                    Log.Error($"Could not access {nameof(AspNetAmbientContext)} for {methodDef}.");
+                }
+            }
+
+            try
+            {
+                // call the original method, catching and rethrowing any unhandled exceptions
+                instrumentedMethod(context);
+            }
+            catch (Exception ex) when (ambientContext?.SetExceptionOnRootSpan(exceptionToGrab.HasValue ? exceptionToGrab.Value : ex) ?? false)
```

Comment:
Nit: I find that code using `GetValueOrDefault()` is easier to write and read than using `HasValue`/`Value`. For example, if you use `GetValueOrDefault()` above when you get `exceptionToGrab`, then `exceptionToGrab` is actually an `Exception` (as you would expect from the name), and you can simplify the call to `SetExceptionOnRootSpan()` here by coalescing:
 ```csharp
... SetExceptionOnRootSpan(exceptionToGrab ?? ex) ...
```

Reply:
Totally agree